### PR TITLE
Fix SM100 DSA backward handling of invalid sparse rows

### PR DESCRIPTION
**Pull request by Jie-Fang**
## Before submitting

- [ ] I agree to license this contribution under the terms of [LICENSE.txt](https://github.com/NVIDIA/cudnn-frontend/blob/develop/LICENSE.txt).
- [x ] I ran `pre-commit run` and committed any formatting changes.
- [ x] I reviewed the Hard Rules in the `AGENTS.md` for each directory this PR touches (see [root AGENTS.md § Reviewing a PR](https://github.com/NVIDIA/cudnn-frontend/blob/develop/AGENTS.md#reviewing-a-pr-human-or-agent)) and my changes comply, or I explain the exception below.
- [ ] I added GitHub labels: one `cat-*`, one or more `area:*` / `op:*`, and one `orig-*` (see [label list](https://github.com/NVIDIA/cudnn-frontend/labels)).

## Affected area

- FE OSS kernels or CuTeDSL

## Summary

- Add robust per-row validity handling to the generic SM100 DSA backward kernel.
- Predicate KV gathers and dKV reductions for negative, out-of-range, padded, and past-`topk_length` indices.
- Carry a compact warp-ballot validity mask into probability recomputation so invalid sparse rows contribute zero gradients.
- Define the limiting behavior for empty rows and infinite attention sinks.
- Compute `O * dO` in FP32 before accumulation for improved dSink accuracy.
- Rebalance load and compute warp register budgets to eliminate spills introduced by the validation path.
- Add L0 regression coverage that explicitly dispatches to the generic SM100 kernel.

## Why

The generic SM100 DSA backward kernel did not consistently exclude invalid sparse rows from all stages of the computation.

Invalid or padded KV rows could still reach probability recomputation, where their physical score slots were exponentiated instead of being treated as masked rows. In addition, some paths did not validate both the lower and upper bounds of active top-k indices before accessing KV memory. Depending on the input, this could produce NaNs, silently corrupted gradients, or out-of-bounds memory accesses.

The sink-normalization code also evaluated indeterminate expressions for empty rows with `LSE == sink == -inf` and for `sink == +inf`. Finally, `O * dO` was multiplied in BF16 before being converted to FP32, reducing dSink accuracy.

The implementation validates each lane-owned sparse index once, packs the result with a warp ballot, predicates the asynchronous KV gather, and reuses the packed validity mask during probability recomputation. This preserves the existing API while making invalid rows contribute no gradients and preventing invalid KV accesses.

## Related issues

Related to #676.

This PR fixes the generic `dsa_bwd_sm100.py` path. The dedicated H16 and H32 SM100 specializations are unchanged and are outside the scope of this PR.

## API and compatibility impact

There are no public API or function-signature changes.

Behavior changes:

- Negative, out-of-range, padded, and past-`topk_length` sparse indices no longer contribute to the result or access KV memory.
- Empty rows and infinite attention sinks produce their mathematically defined limiting gradients instead of NaNs.
- dSink accumulation uses an FP32 `O * dO` product.

Supported GPU, CUDA, cuDNN, and Python versions are unchanged.

Performance was measured on an NVIDIA B200 using BF16, 64 heads, head dimension 512, top-k 512, and `topk_length` enabled:

| Sequence length | Pre-fix unchecked kernel | Fixed kernel | Change |
|---:|---:|---:|---:|
| 4096 | 1.227 ms | 1.271 ms | +3.59% |
| 8192 | 2.491 ms | 2.567 ms | +3.05% |

The remaining overhead is the cost of the required sparse-index validation and validity masking. Compared with the initial safe implementation, the optimized implementation improves performance by approximately 0.8% and removes all observed local-memory spill instructions. Nsight Compute reports zero local-load and local-store instructions for the final kernel.

After merging the latest `develop`, the generic-kernel timing changed by only -0.05% at sequence length 4096 and -0.23% at 8192, which is within measurement noise.

## Testing

Environment:

- NVIDIA B200
- CUDA 13.1
- cuDNN backend 9.20.0
- PyTorch 2.13.0+cu130

Formatting:

```bash
pre-commit run --files \
    python/cudnn/deepseek_sparse_attention/sparse_attention_backward/dsa_bwd_sm100.py \
    test/python/fe_api/dsa/test_DSA_sparse_attention_backward.py
```
Result: all applicable hooks passed.
Targeted issue regressions:
```
cd test/python
python -m pytest -q -m L0 \
    fe_api/dsa/test_DSA_sparse_attention_backward.py::test_DSA_sparse_attention_backward_sm100_masks_invalid_topk_rows \
    fe_api/dsa/test_DSA_sparse_attention_backward.py::test_DSA_sparse_attention_backward_sm100_handles_infinite_sink_limits \
    fe_api/dsa/test_DSA_sparse_attention_backward.py::test_DSA_sparse_attention_backward_sm100_accumulates_odo_in_fp32
```
Result: 8 passed.
Complete DSA backward L0 file:
```
cd test/python
python -m pytest -q -m L0 \
    fe_api/dsa/test_DSA_sparse_attention_backward.py
```
Result: 36 passed, 1 skipped.
Performance benchmark, run for the pre-fix and fixed revisions in alternating order for five measured repetitions after an excluded compile/warm-up run:
```
python benchmark/dsa/benchmark_dsa_sparse_attention_backward.py \
    --seqlens 4096,8192 \
    --topks 512 \
    --nheads 64 \
    --head-dim 512 \
    --warmup 20 \
    --repeat 100 \
    --csv <output.csv>
```

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved sparse-attention backward computation for invalid, out-of-bounds, empty, and padded top-k rows.
  * Corrected gradient handling for infinite sink and log-sum-exp values.
  * Improved numerical accuracy for output-gradient accumulation, including FP32 sink-gradient calculations.
  * Improved handling of boundary head tiles and multi-CTA execution.

* **Tests**
  * Added coverage for deterministic execution, CUDA Graphs, workspace ordering, H128 configurations, stream behavior, and top-k edge cases.

### CODE REVIEW

**Review by jiayus-nvidia — APPROVED** [review]

**Comment by jiayus-nvidia** [comment]
@cudnn-ci-bot run oss
